File: L4/L4_징검다리.cpp
//#include <string>
//#include <vector>
//#include <algorithm>
//
//using namespace std;
//
//int solution(int distance, vector<int> rocks, int n) {
//	sort(rocks.begin(), rocks.end());
//	vector<int> distances;
//	distances.push_back(rocks[0]);
//	for (int i = 1; i < rocks.size();i++) {
//		distances.push_back(rocks[i] - rocks[i - 1]);
//	}
//	distances.push_back(distance - rocks[rocks.size()-1]);
//
//	for (int d : distances) {
//		printf("%d ", d);
//	}
//	printf("\n");
//
//	for (int i = 0; i < n; i++) {
//		auto min = min_element(distances.begin(), distances.end());
//		if (min == distances.end() - 1) *(min - 1) += *min;
//		else *(min + 1) += *min;
//		distances.erase(min);
//
//		for (int d : distances) {
//			printf("%d ", d);
//		}
//		printf("\n");
//
//	}
//
//	return *min_element(distances.begin(), distances.end());
//}

File: Programmers/L4/L4_징검다리.js
function solution(distance, rocks, n) {
    let start = 0, end = 1000000001, answer = 0;
    rocks = rocks.sort((a1, a2) => a1 - a2);
    
    while (start <= end) {
        let mid = Math.floor((start + end) / 2);
        let count = 0, pos = 0;
        for (let rock of rocks) {
            if (mid > rock - pos) count++;
            else pos = rock;
        }
        if (mid > distance - pos) count++;
        
        if (count <= n) {
            answer = Math.max(answer, mid);
            start = mid + 1;
        }
        else end = mid - 1;
    }
    
    return answer;
}